File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/settings/folders/ExcludedFoldersPresenter.java
package com.github.anrimian.musicplayer.ui.settings.folders;

import android.annotation.SuppressLint;

import com.github.anrimian.musicplayer.domain.interactors.library.LibraryFoldersInteractor;
import com.github.anrimian.musicplayer.domain.models.folders.IgnoredFolder;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;

import java.util.List;

import javax.annotation.Nullable;

import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import moxy.MvpPresenter;


public class ExcludedFoldersPresenter extends MvpPresenter<ExcludedFoldersView> {

    private final LibraryFoldersInteractor interactor;
    private final Scheduler uiScheduler;
    private final ErrorParser errorParser;

    private final CompositeDisposable presenterDisposable = new CompositeDisposable();

    @Nullable
    private IgnoredFolder recentlyRemovedFolder;

    public ExcludedFoldersPresenter(LibraryFoldersInteractor interactor,
                                    Scheduler uiScheduler,
                                    ErrorParser errorParser) {
        this.interactor = interactor;
        this.uiScheduler = uiScheduler;
        this.errorParser = errorParser;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        subscribeOnIgnoredFoldersList();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    @SuppressLint("CheckResult")
    void onDeleteFolderClicked(IgnoredFolder folder) {
        //noinspection ResultOfMethodCallIgnored
        interactor.deleteIgnoredFolder(folder)
                .toSingleDefault(folder)
                .observeOn(uiScheduler)
                .subscribe(this::onFolderRemoved, this::onDefaultError);
    }

    void onRestoreRemovedFolderClicked() {
        if (recentlyRemovedFolder == null) {
            return;
        }
        presenterDisposable.add(interactor.addFolderToIgnore(recentlyRemovedFolder)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onFoldersListError));
    }

    private void onFolderRemoved(IgnoredFolder folder) {
        recentlyRemovedFolder = folder;
        getViewState().showRemovedFolderMessage(folder);
    }

    private void subscribeOnIgnoredFoldersList() {
        presenterDisposable.add(interactor.getIgnoredFoldersObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onFoldersListReceived, this::onFoldersListError));
    }

    private void onFoldersListReceived(List<IgnoredFolder> folders) {
        getViewState().showExcludedFoldersList(folders);
        if (folders.isEmpty()) {
            getViewState().showEmptyListState();
        } else {
            getViewState().showListState();
        }
    }

    private void onFoldersListError(Throwable throwable) {
        getViewState().showErrorState(errorParser.parseError(throwable));
    }

    private void onDefaultError(Throwable throwable) {
        getViewState().showErrorMessage(errorParser.parseError(throwable));
    }

}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/settings/folders/ExcludedFoldersPresenter.kt
package com.github.anrimian.musicplayer.ui.settings.folders

import com.github.anrimian.musicplayer.domain.interactors.library.LibraryFoldersInteractor
import com.github.anrimian.musicplayer.domain.models.folders.IgnoredFolder
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.common.mvp.AppPresenter
import io.reactivex.rxjava3.core.Scheduler


class ExcludedFoldersPresenter(
    private val interactor: LibraryFoldersInteractor,
    uiScheduler: Scheduler,
    errorParser: ErrorParser
) : AppPresenter<ExcludedFoldersView>(uiScheduler, errorParser) {

    private var recentlyRemovedFolder: IgnoredFolder? = null

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        subscribeOnIgnoredFoldersList()
    }

    fun onDeleteFolderClicked(folder: IgnoredFolder) {
        interactor.deleteIgnoredFolder(folder)
            .toSingleDefault(folder)
            .launchOnUi(this::onFolderRemoved, viewState::showErrorMessage)
    }

    fun onRestoreRemovedFolderClicked() {
        if (recentlyRemovedFolder == null) {
            return
        }
        interactor.addFolderToIgnore(recentlyRemovedFolder!!)
            .justRunOnUi(viewState::showErrorMessage)
    }

    private fun onFolderRemoved(folder: IgnoredFolder) {
        recentlyRemovedFolder = folder
        viewState.showRemovedFolderMessage(folder)
    }

    private fun subscribeOnIgnoredFoldersList() {
        interactor.getIgnoredFoldersObservable()
            .runOnUi(this::onFoldersListReceived, viewState::showErrorState)
    }

    private fun onFoldersListReceived(folders: List<IgnoredFolder>) {
        viewState.showExcludedFoldersList(folders)
        if (folders.isEmpty()) {
            viewState.showEmptyListState()
        } else {
            viewState.showListState()
        }
    }

}